### ISSUE TITLE
Support Swagger v4

### ISSUE THREAD

**Issue by Hawxy** [micro-elements/MicroElements.Swashbuckle.FluentValidation#20]
Swagger v4 has breaking changes in the way middleware works and this library is currently incompatible.

**Comment by VictorioBerra** [comment]
Error when using the sample:

```
MissingMethodException: Method not found: 'Void Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenOptions.SchemaFilter(System.Object[])'.
Swashbuckle.AspNetCore.Swagger.FluentValidationRulesRegistrator.AddFluentValidationRules(SwaggerGenOptions options)
TestSwashRules.Startup+<>c.<ConfigureServices>b__4_2(SwaggerGenOptions c) in Startup.cs
-
    services.AddSwaggerGen(c =>
    {
        c.SwaggerDoc("v1", new Info { Title = "My API", Version = "v1" });
        // Adds fluent validation rules to swagger
        c.AddFluentValidationRules();
    });
        }
        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
Microsoft.Extensions.Options.ConfigureNamedOptions<TOptions>.Configure(string name, TOptions options)
Microsoft.Extensions.Options.OptionsFactory<TOptions>.Create(string name)
Microsoft.Extensions.Options.OptionsManager<TOptions>+<>c__DisplayClass5_0.<Get>b__0()
System.Lazy<T>.ViaFactory(LazyThreadSafetyMode mode)
System.Lazy<T>.ExecutionAndPublication(LazyHelper executionAndPublication, bool useDefaultConstructor)
System.Lazy<T>.CreateValue()
Microsoft.Extensions.Options.OptionsCache<TOptions>.GetOrAdd(string name, Func<TOptions> createOptions)
Microsoft.Extensions.Options.OptionsManager<TOptions>.Get(string name)
Microsoft.Extensions.Options.OptionsManager<TOptions>.get_Value()
Swashbuckle.AspNetCore.SwaggerGen.ConfigureSchemaRegistryOptions..ctor(IServiceProvider serviceProvider, IOptions<SwaggerGenOptions> swaggerGenOptionsAccessor)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<TArgument, TResult>.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine+<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.GetService(IServiceProvider sp, Type type, Type middleware)
lambda_method(Closure , object , HttpContext , IServiceProvider )
Microsoft.AspNetCore.Builder.UseMiddlewareExtensions+<>c__DisplayClass4_1.<UseMiddleware>b__2(HttpContext context)
Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by bbartels** [comment]
A temporary workaround is to make the calls made in AddFluentValidationRules manually:
```csharp
services.AddSwaggerGen(s =>
{
    s.SwaggerDoc("v1", new Info { Title = "My Api", Version = "v1" });
    //s.AddFluentValidationRules(); // Throws Exception
    //Doing the AddFluentValidationRules' calls manually, no exception
    s.SchemaFilter<FluentValidationRules>();
    s.OperationFilter<FluentValidationOperationFilter>();
});```

**Comment by petriashev** [comment]
Current error with v4 occurs because of namespaces renames in Swashbuckle.AspNetCore.
But there are other errors with scoped services that was worked in my workarounds but was broken in new version.
I can release temp version that fixes naming problem and fix problem with scoped services later.
What do you think?

**Comment by bbartels** [comment]
Maybe release a temporary alpha nuget package to fix the namespace renaming issue and once the scoped services thing is fixed you merge it into a stable one?

**Comment by petriashev** [comment]
Check 2.0.0-beta: https://www.nuget.org/packages/MicroElements.Swashbuckle.FluentValidation/2.0.0-beta.1

**Comment by petriashev** [comment]
I have no workaround for DI problem. I tried to debug issue but I need more time.
I sure that error in Swashbuckle.AspNetCore in new configuration code. Some of code caches IServiceProvider and this leads to error. IServiceProvider should be get on each middleware call (scoped) but internal code uses cached one.

**Comment by VictorioBerra** [comment]
Our code or their code @petriashev ?

**Comment by petriashev** [comment]
> 
> Our code or their code @petriashev ?

Their (Swashbuckle.AspNetCore)
I will try to create fix but I have a little time (deadline on primary work...)

**Comment by VictorioBerra** [comment]
You mean you have a normal job that pays you for work instead of fixing our
problems for free? :P

**Comment by petriashev** [comment]
Right)

**Comment by jondmcelroy** [comment]
Any updates on this? Still appears to not work for me with:
FluentValidation  8.1.2 
FluentValidation.AspNetCore   8.1.2
MicroElements.Swashbuckle.FluentValidation  2.0.0-beta.1
Swashbuckle.AspNetCore 5.0.0-beta

```c#
System.Reflection.ReflectionTypeLoadException: 'Unable to load one or more of the requested types.
Method 'Apply' in type 'MicroElements.Swashbuckle.FluentValidation.FluentValidationOperationFilter' from assembly 'MicroElements.Swashbuckle.FluentValidation, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.
```

**Comment by petriashev** [comment]
Added https://github.com/micro-elements/MicroElements.Swashbuckle.FluentValidation/issues/25 to add support swagger v5

**Comment by IvanFarkas** [comment]
Great.
What about ASP.NET Core 3.0 compatibility since now it's in Preview 2?

**Comment by petriashev** [comment]
@IvanFarkas Is it incompatible with current versions?

**Comment by petriashev** [comment]
closed with release 2.0.0
